From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NONPLAYT <nonplay@bxteam.org>
Date: Sun, 6 Jul 2025 02:38:39 +0300
Subject: [PATCH] C2ME: optimize noise generation

This patch is based on the following mixins:
* "com/ishland/c2me/opts/math/mixin/MixinOctavePerlinNoiseSampler.java"
* "com/ishland/c2me/opts/math/mixin/MixinPerlinNoiseSampler.java"
By: ishland <ishlandmc@yeah.net>
As part of: C2ME (https://github.com/RelativityMC/C2ME-fabric)
Licensed under: MIT (https://opensource.org/licenses/MIT)

Port source: BX-Team/DivineMC@039981b0bf973ea9bee0f2521b2c01d9bb419802
Quantum adaptation: default-off guard, native fallback and correctness fixes.
Earlier project attribution and license retained; outside LICENSE-QUANTUM.

diff --git a/net/minecraft/world/level/levelgen/synth/ImprovedNoise.java b/net/minecraft/world/level/levelgen/synth/ImprovedNoise.java
index ef8f2a0..e8aa601 100644
--- a/net/minecraft/world/level/levelgen/synth/ImprovedNoise.java
+++ b/net/minecraft/world/level/levelgen/synth/ImprovedNoise.java
@@ -11,6 +11,27 @@ public final class ImprovedNoise {
     public final double yo;
     public final double zo;
 
+    // DivineMC start - C2ME: optimize noise generation
+    private static final double[] FLAT_SIMPLEX_GRAD = new double[]{
+        1, 1, 0, 0,
+        -1, 1, 0, 0,
+        1, -1, 0, 0,
+        -1, -1, 0, 0,
+        1, 0, 1, 0,
+        -1, 0, 1, 0,
+        1, 0, -1, 0,
+        -1, 0, -1, 0,
+        0, 1, 1, 0,
+        0, -1, 1, 0,
+        0, 1, -1, 0,
+        0, -1, -1, 0,
+        1, 1, 0, 0,
+        0, -1, 1, 0,
+        -1, 1, 0, 0,
+        0, -1, -1, 0,
+    };
+    // DivineMC end - C2ME: optimize noise generation
+
     public ImprovedNoise(final RandomSource random) {
         this.xo = random.nextDouble() * 256.0;
         this.yo = random.nextDouble() * 256.0;
@@ -83,6 +104,7 @@ public final class ImprovedNoise {
     }
 
     private double sampleAndLerp(final int x, final int y, final int z, final double xr, final double yr, final double zr, final double yrOriginal) {
+        if (dev.quantumspigot.server.QuantumRuntime.worldgen().noiseKernel()) return quantumSampleAndLerp(x, y, z, xr, yr, zr, yrOriginal);
         int x0 = this.p(x);
         int x1 = this.p(x + 1);
         int xy00 = this.p(x0 + y);
@@ -103,6 +125,72 @@ public final class ImprovedNoise {
         return Mth.lerp3(xAlpha, yAlpha, zAlpha, d000, d100, d010, d110, d001, d101, d011, d111);
     }
 
+    private double quantumSampleAndLerp(final int x, final int y, final int z, final double xr, final double yr, final double zr, final double yrOriginal) {
+        // DivineMC start - C2ME: optimize noise generation
+        final int var0 = x & 0xFF;
+        final int var1 = (x + 1) & 0xFF;
+        final int var2 = this.p[var0] & 0xFF;
+        final int var3 = this.p[var1] & 0xFF;
+        final int var4 = (var2 + y) & 0xFF;
+        final int var5 = (var3 + y) & 0xFF;
+        final int var6 = (var2 + y + 1) & 0xFF;
+        final int var7 = (var3 + y + 1) & 0xFF;
+        final int var8 = this.p[var4] & 0xFF;
+        final int var9 = this.p[var5] & 0xFF;
+        final int var10 = this.p[var6] & 0xFF;
+        final int var11 = this.p[var7] & 0xFF;
+
+        final int var12 = (var8 + z) & 0xFF;
+        final int var13 = (var9 + z) & 0xFF;
+        final int var14 = (var10 + z) & 0xFF;
+        final int var15 = (var11 + z) & 0xFF;
+        final int var16 = (var8 + z + 1) & 0xFF;
+        final int var17 = (var9 + z + 1) & 0xFF;
+        final int var18 = (var10 + z + 1) & 0xFF;
+        final int var19 = (var11 + z + 1) & 0xFF;
+        final int var20 = (this.p[var12] & 15) << 2;
+        final int var21 = (this.p[var13] & 15) << 2;
+        final int var22 = (this.p[var14] & 15) << 2;
+        final int var23 = (this.p[var15] & 15) << 2;
+        final int var24 = (this.p[var16] & 15) << 2;
+        final int var25 = (this.p[var17] & 15) << 2;
+        final int var26 = (this.p[var18] & 15) << 2;
+        final int var27 = (this.p[var19] & 15) << 2;
+        final double var60 = xr - 1.0;
+        final double var61 = yr - 1.0;
+        final double var62 = zr - 1.0;
+        final double var87 = FLAT_SIMPLEX_GRAD[(var20) | 0] * xr + FLAT_SIMPLEX_GRAD[(var20) | 1] * yr + FLAT_SIMPLEX_GRAD[(var20) | 2] * zr;
+        final double var88 = FLAT_SIMPLEX_GRAD[(var21) | 0] * var60 + FLAT_SIMPLEX_GRAD[(var21) | 1] * yr + FLAT_SIMPLEX_GRAD[(var21) | 2] * zr;
+        final double var89 = FLAT_SIMPLEX_GRAD[(var22) | 0] * xr + FLAT_SIMPLEX_GRAD[(var22) | 1] * var61 + FLAT_SIMPLEX_GRAD[(var22) | 2] * zr;
+        final double var90 = FLAT_SIMPLEX_GRAD[(var23) | 0] * var60 + FLAT_SIMPLEX_GRAD[(var23) | 1] * var61 + FLAT_SIMPLEX_GRAD[(var23) | 2] * zr;
+        final double var91 = FLAT_SIMPLEX_GRAD[(var24) | 0] * xr + FLAT_SIMPLEX_GRAD[(var24) | 1] * yr + FLAT_SIMPLEX_GRAD[(var24) | 2] * var62;
+        final double var92 = FLAT_SIMPLEX_GRAD[(var25) | 0] * var60 + FLAT_SIMPLEX_GRAD[(var25) | 1] * yr + FLAT_SIMPLEX_GRAD[(var25) | 2] * var62;
+        final double var93 = FLAT_SIMPLEX_GRAD[(var26) | 0] * xr + FLAT_SIMPLEX_GRAD[(var26) | 1] * var61 + FLAT_SIMPLEX_GRAD[(var26) | 2] * var62;
+        final double var94 = FLAT_SIMPLEX_GRAD[(var27) | 0] * var60 + FLAT_SIMPLEX_GRAD[(var27) | 1] * var61 + FLAT_SIMPLEX_GRAD[(var27) | 2] * var62;
+
+        final double var95 = xr * 6.0 - 15.0;
+        final double var96 = yrOriginal * 6.0 - 15.0;
+        final double var97 = zr * 6.0 - 15.0;
+        final double var98 = xr * var95 + 10.0;
+        final double var99 = yrOriginal * var96 + 10.0;
+        final double var100 = zr * var97 + 10.0;
+        final double var101 = xr * xr * xr * var98;
+        final double var102 = yrOriginal * yrOriginal * yrOriginal * var99;
+        final double var103 = zr * zr * zr * var100;
+
+        final double var113 = var87 + var101 * (var88 - var87);
+        final double var114 = var93 + var101 * (var94 - var93);
+        final double var115 = var91 + var101 * (var92 - var91);
+        final double var116 = var89 + var101 * (var90 - var89);
+        final double var117 = var114 - var115;
+        final double var118 = var102 * (var116 - var113);
+        final double var119 = var102 * var117;
+        final double var120 = var113 + var118;
+        final double var121 = var115 + var119;
+        return var120 + (var103 * (var121 - var120));
+        // DivineMC end - C2ME: optimize noise generation
+    }
+
     private double sampleWithDerivative(final int x, final int y, final int z, final double xr, final double yr, final double zr, final double[] derivativeOut) {
         int x0 = this.p(x);
         int x1 = this.p(x + 1);
